src/figma/dev-resources/diff.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  type DevResource,
+  type PostDevResourcesRequestBody,
+  type PutDevResourcesRequestBody,
+} from '@figma/rest-api-spec';
+import yoctoSpinner from 'yocto-spinner';
+import { figmaFileId, resources } from './constants.js';
+
+/**
+ * Performs a diff between the resources returned from Figma's API
+ * against our local resources. The local resources are always
+ * considered our source of truth and will overwrite the existing
+ * resources in Figma. Returns an object with resources grouped by
+ * which items to create, update, and delete.
+ */
+export default (existingResources: DevResource[]) => {
+  const spinner = yoctoSpinner({
+    text: 'Diffing…\n',
+  }).start();
+
+  const toCreate: PostDevResourcesRequestBody['dev_resources'] = [];
+  const toDelete: DevResource[] = [];
+  const toUpdate: PutDevResourcesRequestBody['dev_resources'] = [];
+
+  const localResources = [];
+
+  // Put our hand-rolled list into the format Figma's APIs expect.
+  for (const resource of resources) {
+    for (const nodeId of resource.nodeIds) {
+      localResources.push(
+        {
+          name: 'GitHub',
+          url: resource.urls.github,
+          file_key: figmaFileId,
+          // Node IDs in the URL are separated with a `-`, but with
+          // a `:` via the API.
+          node_id: nodeId.replaceAll('-', ':'),
+        },
+        {
+          name: 'Storybook',
+          url: resource.urls.storybook,
+          file_key: figmaFileId,
+          // Node IDs in the URL are separated with a `-`, but with
+          // a `:` via the API.
+          node_id: nodeId.replaceAll('-', ':'),
+        },
+      );
+    }
+  }
+
+  // A composite key of `node_id:name` is used because each node can
+  // have multiple resources associated with it (GitHub, Storybook).
+  // A node will never have more than one GitHub or Storybook link
+  // though, so this makes the combination unique. Having a unique
```

Comment:
> A node will never have more than one GitHub or Storybook link though, so this makes the combination unique.

This may not _always_ be true, but for now it is. If/When the time comes, we can update.